File: mobile/flutter/lib/features/red_packet/red_packet_service.dart
// Red Packet Service - Flutter Implementation

class RedPacket {
  final String id;
  final String sender;
  final String senderAddress;
  final String token;
  final double amount;
  final int totalCount;
  final int receivedCount;
  final double remainingAmount;
  final int remainingCount;
  final String message;
  final String type; // random or fixed
  final String link;
  final String status;
  final DateTime createTime;
  final DateTime expireTime;

  RedPacket({
    required this.id,
    required this.sender,
    required this.senderAddress,
    required this.token,
    required this.amount,
    required this.totalCount,
    required this.receivedCount,
    required this.remainingAmount,
    required this.remainingCount,
    required this.message,
    required this.type,
    required this.link,
    required this.status,
    required this.createTime,
    required this.expireTime,
  });

  factory RedPacket.create({
    required String sender,
    required String senderAddress,
    required String token,
    required double amount,
    required int totalCount,
    required String type,
    required String message,
  }) {
    final now = DateTime.now();
    final id = 'rp-${now.millisecondsSinceEpoch}';
    return RedPacket(
      id: id,
      sender: sender,
      senderAddress: senderAddress,
      token: token,
      amount: amount,
      totalCount: totalCount,
      receivedCount: 0,
      remainingAmount: amount,
      remainingCount: totalCount,
      message: message,
      type: type,
      link: 'https://tigerwallet.com/redpacket/claim/$id',
      status: 'active',
      createTime: now,
      expireTime: now.add(const Duration(hours: 24)),
    );
  }

  bool get isExpired => DateTime.now().isAfter(expireTime);
  bool get isActive => status == 'active' && !isExpired;
  bool get isCompleted => remainingCount <= 0;
}

class ClaimRecord {
  final String packetId;
  final String claimer;
  final String claimerAddress;
  final double amount;
  final DateTime claimTime;

  ClaimRecord({
    required this.packetId,
    required this.claimer,
    required this.claimerAddress,
    required this.amount,
    required this.claimTime,
  });
}

class RedPacketService {
  static final Map<String, RedPacket> _packets = {};
  static final Map<String, List<ClaimRecord>> _claims = {};

  static RedPacket createPacket({
    required String sender,
    required String senderAddress,
    required String token,
    required double amount,
    required int totalCount,
    required String type,
    required String message,
  }) {
    final packet = RedPacket.create(
      sender: sender,
      senderAddress: senderAddress,
      token: token,
      amount: amount,
      totalCount: totalCount,
      type: type,
      message: message,
    );
    _packets[packet.id] = packet;
    return packet;
  }

  static RedPacket? getPacket(String packetId) {
    return _packets[packetId];
  }

  static RedPacket? getPacketByLink(String link) {
    final id = link.split('/').last;
    return _packets[id];
  }

  static ClaimRecord? claimPacket({
    required String packetId,
    required String claimer,
    required String claimerAddress,
  }) {
    final packet = _packets[packetId];
    if (packet == null) return null;
    if (!packet.isActive) return null;

    double claimAmount;
    if (packet.type == 'random') {
      if (packet.remainingCount == 1) {
        claimAmount = packet.remainingAmount;
      } else {
        claimAmount = (packet.remainingAmount * 2 / packet.remainingCount) * 
            (DateTime.now().millisecond % 100 / 100);
        if (claimAmount > packet.remainingAmount) {
          claimAmount = packet.remainingAmount / 2;
        }
      }
    } else {
      claimAmount = packet.amount / packet.totalCount;
    }

    final claim = ClaimRecord(
      packetId: packetId,
      claimer: claimer,
      claimerAddress: claimerAddress,
      amount: claimAmount,
      claimTime: DateTime.now(),
    );

    _claims[packetId] = [...(_claims[packetId] ?? []), claim];

    // Update packet
    final updatedPacket = RedPacket(
      id: packet.id,
      sender: packet.sender,
      senderAddress: packet.senderAddress,
      token: packet.token,
      amount: packet.amount,
      totalCount: packet.totalCount,
      receivedCount: packet.receivedCount + 1,
      remainingAmount: packet.remainingAmount - claimAmount,
      remainingCount: packet.remainingCount - 1,
      message: packet.message,
      type: packet.type,
      link: packet.link,
      status: packet.remainingCount - 1 <= 0 ? 'completed' : 'active',
      createTime: packet.createTime,
      expireTime: packet.expireTime,
    );
    _packets[packetId] = updatedPacket;

    return claim;
  }

  static List<ClaimRecord> getPacketClaims(String packetId) {
    return _claims[packetId] ?? [];
  }

  static List<RedPacket> getUserPackets(String userId) {
    return _packets.values.where((p) => p.sender == userId).toList();
  }

  static void expirePackets() {
    _packets.forEach((id, packet) {
      if (packet.isActive && packet.isExpired) {
        _packets[id] = RedPacket(
          id: packet.id,
          sender: packet.sender,
          senderAddress: packet.senderAddress,
          token: packet.token,
          amount: packet.amount,
          totalCount: packet.totalCount,
          receivedCount: packet.receivedCount,
          remainingAmount: packet.remainingAmount,
          remainingCount: packet.remainingCount,
          message: packet.message,
          type: packet.type,
          link: packet.link,
          status: 'expired',
          createTime: packet.createTime,
          expireTime: packet.expireTime,
        );
      }
    });
  }
}

File: backend/go/red_packet_service/red_packet_service.go
package red_packet_service

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"sync"
	"time"
)

// ============================================================================
// Types
// ============================================================================

type PacketType string

const (
	PacketTypeRandom PacketType = "random"
	PacketTypeFixed PacketType = "fixed"
)

type RedPacket struct {
	ID              string      `json:"id"`
	Sender          string      `json:"sender"`
	SenderAddress   string      `json:"senderAddress"`
	Token           string      `json:"token"`
	Amount          float64     `json:"amount"`
	TotalCount      int         `json:"totalCount"`
	ReceivedCount   int         `json:"receivedCount"`
	RemainingAmount float64     `json:"remainingAmount"`
	RemainingCount  int         `json:"remainingCount"`
	Message         string      `json:"message"`
	Type            PacketType  `json:"type"`
	Link            string      `json:"link"`
	Status          string      `json:"status"`
	CreateTime      time.Time   `json:"createTime"`
	ExpireTime      time.Time   `json:"expireTime"`
}

type ClaimRecord struct {
	PacketID    string    `json:"packetId"`
	Claimer     string    `json:"claimer"`
	ClaimerAddr string    `json:"claimerAddr"`
	Amount      float64   `json:"amount"`
	ClaimTime   time.Time `json:"claimTime"`
}

type RedPacketService struct {
	mu      sync.RWMutex
	packets map[string]*RedPacket
	claims  map[string][]ClaimRecord
}

func NewRedPacketService() *RedPacketService {
	return &RedPacketService{
		packets: make(map[string]*RedPacket),
		claims:  make(map[string][]ClaimRecord),
	}
}

func (rps *RedPacketService) CreatePacket(sender, senderAddress, token string, amount float64, totalCount int, packetType PacketType, message string) (*RedPacket, error) {
	rps.mu.Lock()
	defer rps.mu.Unlock()

	if amount <= 0 || totalCount <= 0 {
		return nil, fmt.Errorf("invalid amount or count")
	}

	packet := &RedPacket{
		ID:              fmt.Sprintf("rp-%d", time.Now().Unix()),
		Sender:          sender,
		SenderAddress:   senderAddress,
		Token:           token,
		Amount:          amount,
		TotalCount:      totalCount,
		ReceivedCount:   0,
		RemainingAmount: amount,
		RemainingCount:  totalCount,
		Message:         message,
		Type:            packetType,
		Link:            fmt.Sprintf("https://tigerwallet.com/redpacket/claim/%s", fmt.Sprintf("rp-%d", time.Now().Unix())),
		Status:          "active",
		CreateTime:      time.Now(),
		ExpireTime:      time.Now().Add(24 * time.Hour),
	}

	rps.packets[packet.ID] = packet
	rps.packets[packet.Link] = packet

	return packet, nil
}

func (rps *RedPacketService) GetPacket(packetID string) (*RedPacket, error) {
	rps.mu.RLock()
	defer rps.mu.RUnlock()

	packet, ok := rps.packets[packetID]
	if !ok {
		return nil, fmt.Errorf("packet not found: %s", packetID)
	}
	return packet, nil
}

func (rps *RedPacketService) ClaimPacket(packetID, claimer, claimerAddr string) (*ClaimRecord, error) {
	rps.mu.Lock()
	defer rps.mu.Unlock()

	packet, ok := rps.packets[packetID]
	if !ok {
		return nil, fmt.Errorf("packet not found: %s", packetID)
	}

	if packet.Status != "active" {
		return nil, fmt.Errorf("packet is not active")
	}

	if packet.RemainingCount <= 0 {
		return nil, fmt.Errorf("all packets have been claimed")
	}

	var claimAmount float64
	if packet.Type == PacketTypeRandom {
		// Random amount distribution
		if packet.RemainingCount == 1 {
			claimAmount = packet.RemainingAmount
		} else {
			max := packet.RemainingAmount * 2 / float64(packet.RemainingCount)
			claimAmount = rand.Float64() * max
			if claimAmount > packet.RemainingAmount {
				claimAmount = packet.RemainingAmount / 2
			}
		}
	} else {
		// Fixed amount
		claimAmount = packet.Amount / float64(packet.TotalCount)
	}

	claim := &ClaimRecord{
		PacketID:    packetID,
		Claimer:     claimer,
		ClaimerAddr: claimerAddr,
		Amount:      claimAmount,
		ClaimTime:   time.Now(),
	}

	rps.claims[packetID] = append(rps.claims[packetID], *claim)

	packet.ReceivedCount++
	packet.RemainingCount--
	packet.RemainingAmount -= claimAmount

	if packet.RemainingCount <= 0 {
		packet.Status = "completed"
	}

	return claim, nil
}

func (rps *RedPacketService) GetPacketClaims(packetID string) []ClaimRecord {
	rps.mu.RLock()
	defer rps.mu.RUnlock()
	return rps.claims[packetID]
}

func (rps *RedPacketService) GetUserPackets(userID string) []*RedPacket {
	rps.mu.RLock()
	defer rps.mu.RUnlock()

	var packets []*RedPacket
	for _, packet := range rps.packets {
		if packet.Sender == userID {
			packets = append(packets, packet)
		}
	}
	return packets
}

func (rps *RedPacketService) ExpirePackets() {
	rps.mu.Lock()
	defer rps.mu.Unlock()

	now := time.Now()
	for _, packet := range rps.packets {
		if packet.Status == "active" && now.After(packet.ExpireTime) {
			packet.Status = "expired"
		}
	}
}

func (rps *RedPacketService) DeletePacket(packetID string) error {
	rps.mu.Lock()
	defer rps.mu.Unlock()

	if _, ok := rps.packets[packetID]; !ok {
		return fmt.Errorf("packet not found: %s", packetID)
	}

	delete(rps.packets, packetID)
	return nil
}

func (rps *RedPacketService) ToJSON() (string, error) {
	rps.mu.RLock()
	defer rps.mu.RUnlock()

	data := struct {
		Packets map[string]*RedPacket `json:"packets"`
		Claims  map[string][]ClaimRecord `json:"claims"`
	}{
		Packets: rps.packets,
		Claims:  rps.claims,
	}

	jsonData, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return "", err
	}
	return string(jsonData), nil
}
